src/coreclr/tools/Common/Compiler/ObjectWriter/WasmNative.cs

```diff
@@ -0,0 +1,175 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Diagnostics;
+using System.Linq;
+
+namespace ILCompiler.ObjectWriter
+{
+    public enum WasmSectionType
+    {
+        Custom = 0,
+        Type = 1,
+        Import = 2,
+        Function = 3,
+        Table = 4,
+        Memory = 5,
+        Global = 6,
+        Export = 7,
+        Start = 8,
+        Element = 9,
+        Code = 10,
+        Data = 11,
+        DataCount = 12,
+        Tag = 13,
+    }
+
+    public static class DummyValues
+    {
+        // Wasm function signature for (func (params i32) (result i32))
+        public static WasmFuncType CreateWasmFunc_i32_i32()
+        {
+            return new WasmFuncType(
+                paramTypes: new([WasmValueType.I32]),
+                returnTypes: new([WasmValueType.I32])
+            );
+        }
+    }
+
+    public enum WasmValueType
+    {
+        I32 = 0x7F,
+        I64 = 0x7E,
+        F32 = 0x7D,
+        F64 = 0x7C
+    }
+
+    public static class WasmValueTypeExtensions
+    {
+        public static string ToString(this WasmValueType valueType)
+        {
+            return valueType switch
+            {
+                WasmValueType.I32 => "i32",
+                WasmValueType.I64 => "i64",
+                WasmValueType.F32 => "f32",
+                WasmValueType.F64 => "f64",
+                _ => "unknown",
+            };
+        }
+    }
+
+    public struct WasmResultType : IEquatable<WasmResultType>
+    {
+        private readonly WasmValueType[] _types;
+        public readonly Span<WasmValueType> Types => _types;
+
+        public WasmResultType(WasmValueType[] types)
+        {
+            _types = types;
```

Comment:
Can we transparently turn a null types into a 0-element types here to simplify the rest of the code?

Reply:
Yes, definitely.  By transparently do you mean documented?